go/vt/vttablet/tabletserver/repltracker/writer.go

```diff
@@ -309,41 +357,45 @@ func (w *heartbeatWriter) allowNextHeartbeatRequest() {
 func (w *heartbeatWriter) RequestHeartbeats() {
-	if w.onDemandDuration == 0 {
-		// heartbeats are not by demand. Therefore they are just coming in on their own (if enabled)
+	switch w.configType {
+	case HeartbeatConfigTypeAlways:
+		// heartbeats are not by demand. Therefore they are just coming in on their own. No need to lease
+		// heartbeats.
 		return
 	}
 	// In this function we're going to create a timer to activate heartbeats by-demand. Creating a timer has a cost.
 	// Now, this function can be spammed by clients (the lag throttler). We therefore only allow this function to
 	// actually operate once per X seconds (1/4 of onDemandDuration as a reasonable oversampling value):
-	if atomic.LoadInt64(&w.onDemandLastRequestTick) >= atomic.LoadInt64(&w.onDemandRequestTicks) {
-		// Too many requests. We're dropping this one.
+
+	// The writer could be close()d while this function is running and thus the value of the rate limiter could be nil.
+	// We thus use golang atomic here to avoid locking mutexes.
+	rateLimiter := w.safeOnDemandRequestsRateLimiter()
+	if rateLimiter == nil {
 		return
 	}
-	atomic.StoreInt64(&w.onDemandLastRequestTick, atomic.LoadInt64(&w.onDemandRequestTicks))
-
-	// OK, the request passed through.
-
-	w.onDemandMu.Lock()
-	defer w.onDemandMu.Unlock()
-
-	// Now for the actual logic. A client requests heartbeats. If it were only this client, we would
-	// activate heartbeats for the duration of onDemandDuration, and then turn heartbeats off.
-	// However, there may be multiple clients interested in heartbeats, or maybe the same single client
-	// requesting heartbeats again and again. So we keep track of how many _concurrent_ requests we have.
-	// We enable heartbeats as soon as we have a request; we turn heartbeats off once
-	// we have zero concurrent requests
-	w.enableWrites(true)
-	w.concurrentHeartbeatRequests++
-
-	time.AfterFunc(w.onDemandDuration, func() {
+	rateLimiter.Do(func() error {
 		w.onDemandMu.Lock()
 		defer w.onDemandMu.Unlock()
-		w.concurrentHeartbeatRequests--
-		if w.concurrentHeartbeatRequests == 0 {
-			// means there are currently no more clients interested in heartbeats
-			w.enableWrites(false)
-		}
-		w.allowNextHeartbeatRequest()
+
+		// Now for the actual logic. A client requests heartbeats. If it were only this client, we would
+		// activate heartbeats for the duration of onDemandDuration, and then turn heartbeats off.
+		// However, there may be multiple clients interested in heartbeats, or maybe the same single client
+		// requesting heartbeats again and again. So we keep track of how many _concurrent_ requests we have.
+		// We enable heartbeats as soon as we have a request; we turn heartbeats off once
+		// we have zero concurrent requests
+		w.enableWrites()
+		w.concurrentHeartbeatRequests++
+
+		time.AfterFunc(w.onDemandDuration, func() {
+			w.onDemandMu.Lock()
+			defer w.onDemandMu.Unlock()
+			w.concurrentHeartbeatRequests--
+			if w.concurrentHeartbeatRequests == 0 {
```

Comment:
Seems like this is not concurrent safe to me. We can have N goroutines then reading and updating this value, no? I'd use an `atomic.Int64` for this variable. Maybe I'm missing something related the limiter?

Reply:
Please note the mutex just above. This is concurrency safe.

Reply:
Doh! I’m blind 😂